### PR TITLE
Create integration tests for Collection and Singleton handles.

### PR DESCRIPTION
**Pull request by jasonwyatt**
These tests integrate the full Arcs storage stack (without the storage service).

### CODE REVIEW

File: javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    private lateinit var singletonB: SingletonImpl<RawEntity>
+
+    @Before
+    fun setUp() = runBlocking {
+        RamDiskDriverProvider()
+
+        store = Store(STORE_OPTIONS)
+        storageProxy = StorageProxy(store.activate(), CrdtSingleton<RawEntity>())
+
+        singletonA = SingletonImpl("collectionA", storageProxy)
```
Comment (jblebrun) [inline javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt]: minor nit: name should be `singleton*` on these?

**Comment by jblebrun** [comment]
Is it intention that none of the entities tested here contain collections?

File: java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt [hunk 2]
```diff
@@ -120,3 +126,143 @@ sealed class RefModeStoreOutput : Referencable {
...
+            if (firstOp !is CrdtSet.Operation<*>?) {
+                throw CrdtException(
+                    "ReferenceModeStore is managing a Set, Singleton messages not supported"
+                )
+            }
+
+            RefModeOperations(
+                operations.map {
+                    val op = it as CrdtSet.Operation<RawEntity>
+                    when (op) {
```
Comment (jblebrun) [inline java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt]: break out as helper?

File: java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt [hunk 3]
```diff
@@ -120,3 +126,143 @@ sealed class RefModeStoreOutput : Referencable {
...
+        is SingletonType<*> -> {
+            if (firstOp !is CrdtSingleton.Operation<*>?) {
+                throw CrdtException(
+                    "ReferenceModeStore is managing a Singleton, Set messages not supported"
+                )
+            }
+            RefModeOperations(
+                operations.map {
+                    val op = it as CrdtSingleton.Operation<RawEntity>
+                    when (op) {
```
Comment (jblebrun) [inline java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt]: break out as helper?

File: java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt [hunk 4]
```diff
@@ -120,3 +126,143 @@ sealed class RefModeStoreOutput : Referencable {
...
+    is ProxyMessage.Operations<*, *, *> -> sanitizeOperations(storeType)
+    is ProxyMessage.SyncRequest<*, *, *> -> this as RefModeSyncRequest
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun ProxyMessage.ModelUpdate<*, *, *>.sanitizeModelUpdate(
+    storeType: Type
+): RefModeModelUpdate = when (storeType) {
+    is CollectionType<*> -> {
+        if (model !is CrdtSet.Data<*>) {
```
Comment (jblebrun) [inline java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt]: break out as helper method?

File: java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt [hunk 5]
```diff
@@ -120,3 +126,143 @@ sealed class RefModeStoreOutput : Referencable {
...
+        if (model !is RefModeStoreData)
+            RefModeModelUpdate(
+                RefModeStoreData.Set(model as CrdtSet.Data<RawEntity>),
+                id
+            )
+        else this as RefModeModelUpdate
+    }
+    is ReferenceType<*>,
+    is SingletonType<*> -> {
+        if (model !is CrdtSingleton.Data<*>) {
```
Comment (jblebrun) [inline java/arcs/core/storage/referencemode/RefModeCrdtTypes.kt]: break out as helper method?

File: javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt (hunk 1 above)
Comment (jasonwyatt) [reply javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt]: Lol, whoops. Copy-paste has its downsides

**Comment by jasonwyatt** [comment]
> Is it intention that none of the entities tested here contain collections?

It was just easier to write a raw-entity generator that way. I'll add a collection to the Person just to cover that base.  Additionally - I think an evolution of these test cases would be to try to create _two_ storage proxies with _two_ stores, (but for the same storage key)

File: javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            collections = mapOf(
+                "pets" to pets.map { it.toReferencable() }.toSet()
+            )
+        )
+    }
+
+    companion object {
+        private val STORAGE_KEY = ReferenceModeStorageKey(
+            backingKey = RamDiskStorageKey("entities"),
+            storageKey = RamDiskStorageKey("entity_collection")
```
Comment (bayesianmind) [inline javatests/arcs/core/storage/handle/SingletonIntegrationTest.kt]: Consider using a different storage key here, AFAIK we can't guarantee that these two test classes won't be run in parallel on the same host. To the extent this is an example for clients, we can also encourage good naming practices. :)

**Pull request merged**